orbit/pkg/luks/luks_linux.go

```diff
@@ -0,0 +1,221 @@
+//go:build linux
+
+package luks_runner
+
+import (
+	"context"
+	"crypto/rand"
+	"errors"
+	"fmt"
+	"math/big"
+	"regexp"
+	"time"
+
+	"github.com/fleetdm/fleet/v4/orbit/pkg/dialog"
+	"github.com/fleetdm/fleet/v4/orbit/pkg/lvm"
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	"github.com/rs/zerolog/log"
+	"github.com/siderolabs/go-blockdevice/v2/encryption"
+	"github.com/siderolabs/go-blockdevice/v2/encryption/luks"
+)
+
+const (
+	entryDialogTitle     = "Enter disk encryption passphrase"
+	entryDialogText      = "Passphrase:"
+	retryEntryDialogText = "Passphrase incorrect. Please try again."
+	infoFailedTitle      = "Encryption key escrow"
+	infoFailedText       = "Failed to escrow key. Please try again later."
+	infoSuccessTitle     = "Encryption key escrow"
+	infoSuccessText      = "Key escrowed successfully."
+	maxKeySlots          = 8
+)
+
+var ErrKeySlotFull = regexp.MustCompile(`Key slot \d+ is full`)
+
+func (lr *LuksRunner) Run(oc *fleet.OrbitConfig) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	if !oc.Notifications.EscrowLinuxKey {
+		return nil
+	}
+
+	var response LuksResponse
+	key, err := lr.getEscrowKey(ctx)
+	if err != nil {
+		response.Err = err.Error()
+	}
+
+	response.Key = string(key)
+
+	if err := lr.escrower.SendLinuxKeyEscrowResponse(response); err != nil {
+		if err := lr.infoPrompt(ctx, infoFailedTitle, infoFailedText); err != nil {
+			log.Debug().Err(err).Msg("failed to show failed escrow key dialog")
+		}
+		return fmt.Errorf("escrower escrowKey err: %w", err)
+	}
+
+	if response.Err != "" {
+		if err := lr.infoPrompt(ctx, infoFailedTitle, response.Err); err != nil {
+			log.Debug().Err(err).Msg("failed to show failed escrow key dialog")
+		}
+		return fmt.Errorf("error getting linux escrow key: %s", response.Err)
+	}
+
+	// Show success dialog
+	if err := lr.infoPrompt(ctx, infoSuccessTitle, infoSuccessText); err != nil {
+		log.Debug().Err(err).Msg("failed to show success escrow key dialog")
+	}
+
+	return nil
+}
+
+func (lr *LuksRunner) getEscrowKey(ctx context.Context) ([]byte, error) {
+	devicePath, err := lvm.FindRootDisk()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to find LUKS Root Partition: %w", err)
+	}
+
+	device := luks.New(luks.AESXTSPlain64Cipher)
+
+	// Prompt user for existing LUKS passphrase
+	passphrase, err := lr.entryPrompt(ctx, entryDialogTitle, entryDialogText)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to show passphrase entry prompt: %w", err)
+	}
+
+	// Validate the passphrase
+	for {
```

Comment:
purposeful loop as product requirement is to continue to reprompt on invalid password attempts.  A 1 minute timeout is passed into zenity which should protect a leaked go routine.